### PR TITLE
Rewrite policy in Rust

### LINKED ISSUE #15: Rewrite the policy using Rust/Go
Currently this policy is written using Swift, more specifically with the experimental project SwiftWasm compiler.

Unfortunately the compiled code has a random behavior at runtime. It turns out that sometimes accessing a dictionary by key will not find the associated value.

This makes the execution of the policy unpredictable.

We should then rewrite this policy using a more robust programming language: Rust or Go.

## Acceptance criteria

* [x] The old codebase is left inside of this repository inside of a dedicated branch (like `swift-implementation`). Just for historical reasons
* [x] The policy is rewritten using Rust/Go
* [ ] A new version of the policy is tagged and released 
* [ ] Ensure the `artifacthub-pkg.yml` is automatically generated

### PR DESCRIPTION
**Pull request by jvanz**
## Description

Rewrite policy in Rust.

Fix #15

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+                    validation_request,
+                    &pod_spec,
+                    format!("runtime '{}' is reserved", runtime_class_name),
+                );
+            }
+        }
+
+        // The object does not define the runtime. Therefore, the default one will
+        // be used. Update the object to use the fallback runtime.
+
```
Comment (flavio) [inline src/lib.rs:58]: nit, remove the empty line

File: src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+        {
+            if default_runtime_reserved {
+                return mutate_or_reject(
+                    validation_request,
+                    &pod_spec,
+                    "Usage of the default runtime is reserved".to_string(),
+                );
+            }
+        }
+        if let Some(ref fallback_runtime) = validation_request.settings.fallback_runtime {
```
Comment (flavio) [inline src/lib.rs:69]: nit:

```suggestion
        // The default runtime is not reserved, but a fallback is specified. Therefore
        // the Pod is mutated to use the fallback runtime.
        if let Some(ref fallback_runtime) = validation_request.settings.fallback_runtime {
```

File: src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+        #[allow(dead_code)]
+        pub fn mutate_pod_spec_from_request<T: std::default::Default + 'static>(
+            _validation_request: ValidationRequest<T>,
+            _pod_spec: PodSpec,
+        ) -> wapc_guest::CallResult {
+            Err(Box::new(errors::new(errors::ErrorKind::HostError(
+                Vec::new(),
+            ))))
+        }
+    }
```
Comment (flavio) [inline src/lib.rs:126]: All this code can be removed

File: Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+[lib]
+crate-type = ["cdylib"]
+
+[dependencies]
+anyhow = "1.0"
+k8s-openapi-derive = "0.18.0"
+k8s-openapi = { version = "0.18.0", features = ["schemars", "v1_26"] }
+kubewarden-policy-sdk = "0.9.2"
+lazy_static = "1.4"
```
Comment (flavio) [inline Cargo.toml:17]: This can be dropped, we are not using it

File: src/settings.rs [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+use serde::{Deserialize, Serialize};
+//use slog::info;
+
+// Describe the settings your policy expects when
+// loaded by the policy server.
+#[derive(Serialize, Deserialize, Default, Debug)]
+#[serde(default)]
+#[serde(rename_all = "camelCase")]
+pub(crate) struct Settings {
+    pub reserved_runtimes: Vec<String>,
```
Comment (flavio) [inline src/settings.rs:12]: I would use a `Set` instead, this provides better performance when doing lookups (like we do inside of validate)

File: src/settings.rs [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        };
+        let expected_json = json!({
+            "reservedRuntimes": ["runc"],
+            "fallbackRuntime": "gvisor",
+            "defaultRuntimeReserved": true,
+        });
+        let json_value = serde_json::to_value(settings)?;
+        assert_eq!(json_value, expected_json);
+        Ok(())
+    }
```
Comment (flavio) [inline src/settings.rs:55]: I think we can remove this test. I appreciate the concerns about breaking the settings between the Rust and the Swift version, but this is a one time thing (the swift -> rust migration)

File: src/settings.rs [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+    #[rstest]
+    #[case("runc", "runc", false)]
+    #[case("runc", "gvisor", true)]
+    fn fallback_runtime_cannot_be_reserved_runtimes_test(
```
Comment (flavio) [inline src/settings.rs:60]: nit:

```suggestion
    fn validation(
```

**Review by flavio — CHANGES_REQUESTED** [review]
Overall great. There's one thing I would like to be changed: getting rid of the mocking library. I'm not a big fan of using this library; it makes things harder to grasp, introduces dependencies and code changes via conditional compilation flags.
There are circumstances where mocking is the only way, like when dealing with host capabilities, but for a policy like that this is not needed.

File: src/settings.rs (hunk 6 above)
Comment (jvanz) [reply src/settings.rs:55]: Yes, I've added this during development and forgot to remove it. Thanks for the heads up

**Review by viccuad — APPROVED** [review]
LGTM!

The 2 last sections on README.md ("obtain policy" and "Using the policy") are outdated and can be removed too.

File: Cargo.toml [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[lib]
+crate-type = ["cdylib"]
+
+[dependencies]
+anyhow = "1.0"
+k8s-openapi-derive = "0.18.0"
```
Comment (flavio) [inline Cargo.toml:14]: I think this can be removed. This is needed only when defining a CRD in Rust

File: Cargo.toml [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+k8s-openapi = { version = "0.18.0", features = ["v1_26"] }
+kubewarden-policy-sdk = "0.9.2"
```
Comment (flavio) [inline Cargo.toml:16]: Bump to 0.9.4 please

**Review by flavio — APPROVED** [review]
I left one last comment, once this gets addressed the PR can be merged

**Pull request merged**